File: src/main/java/priv/bajdcc/LALR1/syntax/ISyntaxComponentVisitor.java
package priv.bajdcc.LALR1.syntax;

import priv.bajdcc.LALR1.syntax.exp.*;
import priv.bajdcc.util.VisitBag;

/**
 * 基于文法组件的访问接口（Visitor模式）
 *
 * @author bajdcc
 */
public interface ISyntaxComponentVisitor {

	void visitBegin(TokenExp node, VisitBag bag);

	void visitBegin(RuleExp node, VisitBag bag);

	void visitBegin(SequenceExp node, VisitBag bag);

	void visitBegin(BranchExp node, VisitBag bag);

	void visitBegin(OptionExp node, VisitBag bag);

	void visitBegin(PropertyExp node, VisitBag bag);

	void visitEnd(TokenExp node);

	void visitEnd(RuleExp node);

	void visitEnd(SequenceExp node);

	void visitEnd(BranchExp node);

	void visitEnd(OptionExp node);

	void visitEnd(PropertyExp node);
}

File: src/main/kotlin/com/bajdcc/OP/syntax/ISyntaxComponentVisitor.kt
package com.bajdcc.OP.syntax

import com.bajdcc.OP.syntax.exp.BranchExp
import com.bajdcc.OP.syntax.exp.RuleExp
import com.bajdcc.OP.syntax.exp.SequenceExp
import com.bajdcc.OP.syntax.exp.TokenExp
import com.bajdcc.util.VisitBag

/**
 * 基于文法组件的访问接口（Visitor模式）
 *
 * @author bajdcc
 */
interface ISyntaxComponentVisitor {

    fun visitBegin(node: TokenExp, bag: VisitBag)

    fun visitBegin(node: RuleExp, bag: VisitBag)

    fun visitBegin(node: SequenceExp, bag: VisitBag)

    fun visitBegin(node: BranchExp, bag: VisitBag)

    fun visitEnd(node: TokenExp)

    fun visitEnd(node: RuleExp)

    fun visitEnd(node: SequenceExp)

    fun visitEnd(node: BranchExp)

}
